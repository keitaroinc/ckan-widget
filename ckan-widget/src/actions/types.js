export const PACKAGE_SEARCH = 'package_search';
export const ERROR = 'error';
