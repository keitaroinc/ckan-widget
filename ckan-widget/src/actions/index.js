import axios from 'axios';

import {
  PACKAGE_SEARCH,
  ERROR,
  ADD_ORG,
  ADD_GROUP,
  ADD_TAG
} from './types';


export function packageSearch({
  ckanAPI = 'https://trouver.datasud.fr', q = '*:*', rows = 10, start = 0, page = 0, sort = 'score desc, metadata_modified desc', fq='',
  firstCall=false} = {}) {
  const facetsParams = `facet.field=${JSON.stringify(['organization',
                                                      'groups',
                                                      'tags',
                                                      'res_format',
                                                      'license_id',
                                                      'datatype',
                                                      'support',
                                                      'update_frequency'])}`;

  return (dispatch) => {
    axios.get(`${ckanAPI}/api/action/package_search?q=${q}&rows=${rows}&start=${start}&${facetsParams}&sort=${sort}&fq=${fq}`)
      .then((response) => {
        dispatch({
          type: PACKAGE_SEARCH,
          search: q !== '*:*' ? q : '',
          page,
          rows,
          sort,
          facet_search : fq,
          payload: response.data,
          ckanAPI: ckanAPI,
          firstCall
        });
      })
      .catch((error) => {
        let errorMsg = '';
        if (error.response !== undefined) {
          errorMsg = error.response.data;
        } else {
          errorMsg = error.message;
        }

        dispatch({
          type: ERROR,
          payload: errorMsg,
        });
      });
  };
}

export function packageAutocomplete({ ckanAPI = 'https://trouver.datasud.fr', q = '', limit = 10 } = {}) {
  return axios.get(`${ckanAPI}/api/action/package_autocomplete?q=${q}&limit=${limit}`)
    .then(response => response.data);
}

export function organizationShow({ ckanAPI = 'https://trouver.datasud.fr', id = '' } = {}) {
  return (dispatch) => {
    axios.get(`${ckanAPI}/api/action/organization_show?id=${id}&include_dataset_count=false&include_extras=false&include_users=false&include_groups=false`)
    .then((response) => {
      dispatch({
        type: ADD_ORG,
        payload: response.data,
      });
    })
    .catch((error) => {
      let errorMsg = '';
      if (error.response !== undefined) {
        errorMsg = error.response.data.error.message;
      } else {
        errorMsg = error.message;
      }

      dispatch({
        type: ERROR,
        payload: errorMsg,
      });
    });
  };
}

export function groupShow({ ckanAPI = 'https://trouver.datasud.fr', id = '' } = {}) {
  return (dispatch) => {
    axios.get(`${ckanAPI}/api/action/group_show?id=${id}&include_dataset_count=false&include_extras=false&include_users=false&include_groups=false`)
      .then((response) => {
        dispatch({
          type: ADD_GROUP,
          payload: response.data,
        });
      })
      .catch((error) => {
        let errorMsg = '';
        if (error.response !== undefined) {
          errorMsg = error.response.data.error.message;
        } else {
          errorMsg = error.message;
        }

        dispatch({
          type: ERROR,
          payload: errorMsg,
        });
      });
  };
}

export function tagShow({ ckanAPI = 'https://trouver.datasud.fr', id = '' } = {}) {
  return (dispatch) => {
    axios.get(`${ckanAPI}/api/action/tag_show?id=${id}&include_dataset_count=false&include_extras=false&include_users=false&include_groups=false`)
      .then((response) => {
        dispatch({
          type: ADD_TAG,
          payload: response.data,
        });
      })
      .catch((error) => {
        let errorMsg = '';
        if (error.response !== undefined) {
          errorMsg = error.response.data.error.message;
        } else {
          errorMsg = error.message;
        }

        dispatch({
          type: ERROR,
          payload: errorMsg,
        });
      });
  }
}
