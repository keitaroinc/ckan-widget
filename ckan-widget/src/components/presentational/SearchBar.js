import React, { Component } from 'react'
import Autocomplete from 'react-autocomplete'
import { packageAutocomplete } from '../../actions'


class SearchBar extends Component{
    constructor() {
        super()
        this.state = {
            value: '',
            isOpen: false,
            suggestions: []
        }
    }

    componentDidMount() {
        packageAutocomplete().then(data => {
            const suggestions = data.result
            if (Array.isArray(suggestions) && suggestions.length) {
                this.setState({ suggestions })
            }
        })
    }

    handleOnChange = (event, value) => {
        this.setState({ value })
        packageAutocomplete({ q: value }).then(data => {
            const suggestions = data.result
            if (Array.isArray(suggestions) && suggestions.length) {
                this.setState({ suggestions })
            }
        })
        this.props.handleInputChange(event, value)
    }

    handleOnSubmit = event => {
        const value = this.state.value
        this.props.handleInputChange(event, value)
    }

    render(){
        return(
            <form className="search-item" onSubmit={this.handleOnSubmit}>
                <Autocomplete
                    inputProps={{
                        id: 'datasets-autocomplete',
                        placeholder: 'Search datasets'
                    }}
                    wrapperStyle={{ position: 'relative', display: 'inline-block' }}l
                    value={this.state.value}
                    items={this.state.suggestions}
                    open={( this.state.value.length >= 1 && this.state.isOpen )}
                    onMenuVisibilityChange={isOpen => this.setState({ isOpen })}
                    autoHighlight={false}
                    getItemValue={(item) => item.name}
                    onSelect={(value, item) => {
                        this.setState({ value, suggestions: [item] })
                        this.props.handleInputChange(null, value)
                    }}
                    onChange={this.handleOnChange}
                    renderMenu={children => (
                        <div className="menu">
                            {children}
                        </div>
                    )}
                    renderItem={(item, isHighlighted) => (
                        <div className={`item ${isHighlighted ? 'item-highlighted' : ''}`} key={item.match_displayed}>
                            {item.match_displayed}
                        </div>
                    )}
                />
            </form>
        )
    }
}

export default SearchBar
