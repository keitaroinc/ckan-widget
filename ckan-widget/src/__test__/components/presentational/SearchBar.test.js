import React from 'react';
import { shallow } from 'enzyme';
import SearchBar from '../../../components/presentational/SearchBar';
import * as actions from '../../../actions'

describe('SearchBar', () => {
  it('should be defined', () => {
    expect(SearchBar).toBeDefined();
  });

  it('should render correctly', () => {
    actions.packageAutocomplete = () =>{
      return new Promise(
        function (resolve, reject) {
          const resp = {
            data: ['Samsung']
          };
          resolve(resp); // fulfilled
        }
      );
    }


    const tree = shallow(
      <SearchBar />
    );
    expect(tree).toMatchSnapshot();
  });
});