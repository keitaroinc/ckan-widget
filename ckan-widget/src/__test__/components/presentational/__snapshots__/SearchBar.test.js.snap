// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBar />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Autocomplete
        autoHighlight={false}
        getItemValue={[Function]}
        inputProps={
                Object {
                        "className": "form-control form-control-lg border-0",
                        "id": "datasets-autocomplete",
                        "placeholder": "Search datasets",
                      }
        }
        isItemSelectable={[Function]}
        items={Array []}
        menuStyle={
                Object {
                        "background": "rgba(255, 255, 255, 0.9)",
                        "borderRadius": "3px",
                        "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
                        "fontSize": "90%",
                        "maxHeight": "50%",
                        "overflow": "auto",
                        "padding": "2px 0",
                        "position": "fixed",
                      }
        }
        onChange={[Function]}
        onMenuVisibilityChange={[Function]}
        onSelect={[Function]}
        open={false}
        renderInput={[Function]}
        renderItem={[Function]}
        renderMenu={[Function]}
        selectOnBlur={false}
        value=""
        wrapperProps={Object {}}
        wrapperStyle={
                Object {
                        "position": "relative",
                      }
        }
/>,
      "className": "my-5",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "autoHighlight": false,
        "getItemValue": [Function],
        "inputProps": Object {
          "className": "form-control form-control-lg border-0",
          "id": "datasets-autocomplete",
          "placeholder": "Search datasets",
        },
        "isItemSelectable": [Function],
        "items": Array [],
        "menuStyle": Object {
          "background": "rgba(255, 255, 255, 0.9)",
          "borderRadius": "3px",
          "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
          "fontSize": "90%",
          "maxHeight": "50%",
          "overflow": "auto",
          "padding": "2px 0",
          "position": "fixed",
        },
        "onChange": [Function],
        "onMenuVisibilityChange": [Function],
        "onSelect": [Function],
        "open": false,
        "renderInput": [Function],
        "renderItem": [Function],
        "renderMenu": [Function],
        "selectOnBlur": false,
        "value": "",
        "wrapperProps": Object {},
        "wrapperStyle": Object {
          "position": "relative",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Autocomplete
          autoHighlight={false}
          getItemValue={[Function]}
          inputProps={
                    Object {
                              "className": "form-control form-control-lg border-0",
                              "id": "datasets-autocomplete",
                              "placeholder": "Search datasets",
                            }
          }
          isItemSelectable={[Function]}
          items={Array []}
          menuStyle={
                    Object {
                              "background": "rgba(255, 255, 255, 0.9)",
                              "borderRadius": "3px",
                              "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
                              "fontSize": "90%",
                              "maxHeight": "50%",
                              "overflow": "auto",
                              "padding": "2px 0",
                              "position": "fixed",
                            }
          }
          onChange={[Function]}
          onMenuVisibilityChange={[Function]}
          onSelect={[Function]}
          open={false}
          renderInput={[Function]}
          renderItem={[Function]}
          renderMenu={[Function]}
          selectOnBlur={false}
          value=""
          wrapperProps={Object {}}
          wrapperStyle={
                    Object {
                              "position": "relative",
                            }
          }
/>,
        "className": "my-5",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoHighlight": false,
          "getItemValue": [Function],
          "inputProps": Object {
            "className": "form-control form-control-lg border-0",
            "id": "datasets-autocomplete",
            "placeholder": "Search datasets",
          },
          "isItemSelectable": [Function],
          "items": Array [],
          "menuStyle": Object {
            "background": "rgba(255, 255, 255, 0.9)",
            "borderRadius": "3px",
            "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
            "fontSize": "90%",
            "maxHeight": "50%",
            "overflow": "auto",
            "padding": "2px 0",
            "position": "fixed",
          },
          "onChange": [Function],
          "onMenuVisibilityChange": [Function],
          "onSelect": [Function],
          "open": false,
          "renderInput": [Function],
          "renderItem": [Function],
          "renderMenu": [Function],
          "selectOnBlur": false,
          "value": "",
          "wrapperProps": Object {},
          "wrapperStyle": Object {
            "position": "relative",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
